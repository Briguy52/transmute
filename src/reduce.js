import curry from './curry';
import _reduce from './internal/_reduce';

/**
 * Transform the contents of `subject` to `into` by applying `operation` to each
 * item.
 *
 * @param  {any} into
 * @param  {Function} operation
 * @param  {Iterable} subject   [description]
 * @return {Iterable}
 */
export default curry(_reduce);
