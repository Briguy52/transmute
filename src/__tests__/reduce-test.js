import { List, Map, Seq, Set } from "immutable";
import reduce from "../reduce";

describe("transmute/reduce", () => {
  it("reduces an Array", () => {
    const reducer = jest.fn((acc, n) => acc.push(n + 1));
    expect(reduce(List(), reducer, [1, 2, 3])).toMatchSnapshot();
    expect(reducer.mock.calls.length).toBe(3);
    expect(reducer.mock.calls[0]).toEqual([List(), 1, 0, [1, 2, 3]]);
    expect(reducer.mock.calls[1]).toEqual([List.of(2), 2, 1, [1, 2, 3]]);
    expect(reducer.mock.calls[2]).toEqual([List.of(2, 3), 3, 2, [1, 2, 3]]);
  });

  it("reduces a List", () => {
    expect(
      reduce(List(), (acc, n) => acc.push(n + 1), List.of(1, 2, 3))
    ).toMatchSnapshot();
  });

  it("reduces a Map", () => {
    expect(
      reduce(
        Map(),
        (acc, n, k) => acc.set(k, n + 1),
        Map({ one: 1, two: 2, three: 3 })
      )
    ).toMatchSnapshot();
  });

  it("reduces an Object", () => {
    expect(
      reduce(List(), (acc, n) => acc.push(n + 1), { one: 1, two: 2, three: 3 })
    ).toMatchSnapshot();
  });

  it("reduces a Seq", () => {
    expect(
      reduce(Seq(), (acc, n) => acc.concat([n + 1]), Seq.of(1, 2, 3))
    ).toMatchSnapshot();
  });

  it("reduces a Set", () => {
    expect(
      reduce(Set(), (acc, n) => acc.add(n + 1), Set.of(1, 2, 3))
    ).toMatchSnapshot();
  });

  it("reduces to a primitive", () => {
    expect(reduce(0)((acc, n) => acc + n)(List.of(1, 2, 3))).toMatchSnapshot();
  });
});
