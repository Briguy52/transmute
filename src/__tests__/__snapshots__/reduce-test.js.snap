// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/reduce reduces a List 1`] = `
Immutable.List [
  2,
  3,
  4,
]
`;

exports[`transmute/reduce reduces a Map 1`] = `
Immutable.Map {
  "one": 2,
  "two": 3,
  "three": 4,
}
`;

exports[`transmute/reduce reduces a Seq 1`] = `Immutable.Seq [â€¦]`;

exports[`transmute/reduce reduces a Set 1`] = `
Immutable.Set [
  2,
  3,
  4,
]
`;

exports[`transmute/reduce reduces an Array 1`] = `
Immutable.List [
  2,
  3,
  4,
]
`;

exports[`transmute/reduce reduces an Object 1`] = `
Immutable.List [
  2,
  3,
  4,
]
`;

exports[`transmute/reduce reduces to a primitive 1`] = `6`;
